### Data Exploration

In [1]:
#general imports that we will need will almost always use - it is a good practice to import all libraries at the beginning of the notebook or script
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
import os
import json

# data partition
from sklearn.model_selection import train_test_split

#filter methods
# spearman 
# chi-square
import scipy.stats as stats
from scipy.stats import chi2_contingency

#wrapper methods
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.feature_selection import RFE


# embedded methods
from sklearn.linear_model import LassoCV

# ignore warnings
import warnings
warnings.filterwarnings('ignore')

#set random seed for reproducibility
RSEED = 42
np.random.seed(RSEED)

In [2]:
# Load the data paths
data_dir = "../data/"

# Load the raw data into a pandas dataframe

test = pd.read_csv(os.path.join(data_dir, "test.csv"))
train = pd.read_csv(os.path.join(data_dir, "train.csv"))
sample = pd.read_csv(os.path.join(data_dir, "sample_submission.csv"))

In [3]:
# Shapes (rows, columns)
print("Train shape:", train.shape)
print("Test shape:", test.shape)
print("Sample submission shape:", sample.shape)

# Column names
print("\nTrain columns:", train.columns.tolist())
print("\nTest columns:", test.columns.tolist())

# First 5 rows (peek at data)
display(train.head())
display(test.head())
display(sample.head())

# Info (dtypes, non-null counts)
print("\nTrain info:")
print(train.info())
print("\nTest info:")
print(test.info())


Train shape: (75973, 14)
Test shape: (32567, 13)
Sample submission shape: (32567, 2)

Train columns: ['carID', 'Brand', 'model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'tax', 'mpg', 'engineSize', 'paintQuality%', 'previousOwners', 'hasDamage']

Test columns: ['carID', 'Brand', 'model', 'year', 'transmission', 'mileage', 'fuelType', 'tax', 'mpg', 'engineSize', 'paintQuality%', 'previousOwners', 'hasDamage']


,carID,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
0,69512,VW,Golf,2016.0,22290,Semi-Auto,28421.0,Petrol,NaN,11.417268,2.0,63.0,4.000000,0.0
1,53000,Toyota,Yaris,2019.0,13790,Manual,4589.0,Petrol,145.0,47.900000,1.5,50.0,1.000000,0.0
2,6366,Audi,Q2,2019.0,24990,Semi-Auto,3624.0,Petrol,145.0,40.900000,1.5,56.0,4.000000,0.0
3,29021,Ford,FIESTA,2018.0,12500,anual,9102.0,Petrol,145.0,65.700000,1.0,50.0,-2.340306,0.0
4,10062,BMW,2 Series,2019.0,22995,Manual,1000.0,Petrol,145.0,42.800000,1.5,97.0,3.000000,0.0


,carID,Brand,model,year,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
0,89856,Hyundai,I30,2022.878006,Automatic,30700.000000,petrol,205.0,41.5,1.6,61.0,3.0,0.0
1,106581,VW,Tiguan,2017.000000,Semi-Auto,-48190.655673,Petrol,150.0,38.2,2.0,60.0,2.0,0.0
2,80886,BMW,2 Series,2016.000000,Automatic,36792.000000,Petrol,125.0,51.4,1.5,94.0,2.0,0.0
3,100174,Opel,Grandland X,2019.000000,Manual,5533.000000,Petrol,145.0,44.1,1.2,77.0,1.0,0.0
4,81376,BMW,1 Series,2019.000000,Semi-Auto,9058.000000,Diesel,150.0,51.4,2.0,45.0,4.0,0.0


,carID,price
0,89856,851000
1,106581,514000
2,80886,323000
3,100174,921000
4,81376,620000



Train info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75973 entries, 0 to 75972
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   carID           75973 non-null  int64  
 1   Brand           74452 non-null  object 
 2   model           74456 non-null  object 
 3   year            74482 non-null  float64
 4   price           75973 non-null  int64  
 5   transmission    74451 non-null  object 
 6   mileage         74510 non-null  float64
 7   fuelType        74462 non-null  object 
 8   tax             68069 non-null  float64
 9   mpg             68047 non-null  float64
 10  engineSize      74457 non-null  float64
 11  paintQuality%   74449 non-null  float64
 12  previousOwners  74423 non-null  float64
 13  hasDamage       74425 non-null  float64
dtypes: float64(8), int64(2), object(4)
memory usage: 8.1+ MB
None

Test info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32567 entries, 0 to 32566
Dat

In [4]:
# Description of train
train.describe().T

,count,mean,std,min,25%,50%,75%,max
carID,75973.0,37986.000000,21931.660338,0.000000,18993.00,37986.0,56979.0,75972.000000
year,74482.0,2017.096611,2.208704,1970.000000,2016.00,2017.0,2019.0,2024.121759
price,75973.0,16881.889553,9736.926322,450.000000,10200.00,14699.0,20950.0,159999.000000
mileage,74510.0,23004.184088,22129.788366,-58540.574478,7423.25,17300.0,32427.5,323000.000000
tax,68069.0,120.329078,65.521176,-91.121630,125.00,145.0,145.0,580.000000
mpg,68047.0,55.152666,16.497837,-43.421768,46.30,54.3,62.8,470.800000
engineSize,74457.0,1.660136,0.573462,-0.103493,1.20,1.6,2.0,6.600000
paintQuality%,74449.0,64.590667,21.021065,1.638913,47.00,65.0,82.0,125.594308
previousOwners,74423.0,1.994580,1.472981,-2.345650,1.00,2.0,3.0,6.258371
hasDamage,74425.0,0.000000,0.000000,0.000000,0.00,0.0,0.0,0.000000


In [5]:
# Split the training data into features and target variable (price)
X = train.drop(columns=['price'])
y = train['price']

### We check for missing values and wrong values in the column Brand

In [ ]:
# Look at the Missing values in x and y
missing_values_x = X.isnull().sum()
missing_values_y = y.isnull().sum()

print("\nMissing values in X:\n", missing_values_x[missing_values_x > 0])
print("\nMissing values in y:\n", missing_values_y[missing_values_y > 0])

In [ ]:
# unique brands in the column brand
unique_brands = X['Brand'].unique()
print("\nUnique brands in 'Brand':\n", unique_brands)

### First we fix the brand names and afterwards we will map the brands from the model names

In [ ]:
# Load the JSON mapping
with open("../mapping/brandname_mapping.json", "r", encoding="utf-8") as f:
    CANON = json.load(f)

In [ ]:
import re

# Normalization function
def norm_brand(brand):
    if pd.isna(brand):
        return brand
    brand = brand.strip().lower()
    brand = re.sub(r'[.,-_]', ' ', brand)
    brand = ' '.join(brand.split())  # Remove extra spaces
    return brand

In [ ]:
# Cleaning function using imported JSON
def clean_brand(col: pd.Series) -> pd.Series:
    nb = col.apply(norm_brand)
    mapped = nb.map(CANON)
    return mapped

In [ ]:
# apply normalization and canonical mapping
X['Brand'] = X['Brand'].apply(norm_brand).map(CANON)

In [ ]:
# unique brands in the column brand
unique_brands = X['Brand'].unique()
print("\nUnique brands in 'Brand':\n", unique_brands)

In [ ]:
# Show the entries with missing values in brand
missing_brand_entries = X[X['Brand'].isnull()]
display(missing_brand_entries)

- We fixed wrong names in the column Brand but we have missing values. We will try to fill them by looking at the column model.
But there are also some wrong values in the column model. We will try to fix them first


In [ ]:
# Count the unique values of model in x
unique_models = X['model'].value_counts()
print("\nCounts of unique values in 'model' for entries with missing 'Brand':\n", unique_models)

# save the unique_models to a json file
unique_models.to_json("../mapping/unique_model_counts.json")

- Canonical Mapping and Normalization for model names

In [ ]:
import re
import pandas as pd
import numpy as np
import json

def norm_model(x: object) -> str | float:
    """Lowercase, keep [a-z0-9+], strip spaces/hyphens/underscores."""
    if pd.isna(x):
        return np.nan
    s = str(x).lower()
    s = re.sub(r'[\s\-_]+', '', s)           # drop spaces, hyphens, underscores
    s = re.sub(r'[^a-z0-9\+]', '', s)        # keep letters, digits, plus
    return s or np.nan


In [ ]:
# Load mapping JSON
with open("../mapping/modelname_mapping.json", "r", encoding="utf-8") as f:
    MM = json.load(f)

ALIASES = MM.get("aliases", {})
REGEX_RULES = MM.get("regex_rules", [])

In [ ]:
def apply_regex_rules(norm_key: str) -> str:
    for rule in REGEX_RULES:
        pat = rule["pattern"]
        rep = rule["replace"]
        norm_key = re.sub(pat, rep, norm_key)
    return norm_key

In [ ]:
def clean_model(series: pd.Series) -> pd.Series:
    normed = series.apply(norm_model)
    normed = normed.apply(lambda x: apply_regex_rules(x) if isinstance(x, str) else x)
    return normed.map(ALIASES)

In [ ]:
# Apply the clean model function to the 'model' column
X['model'] = clean_model(X['model'])

# Apply the apply_regex_rules function to the 'model' column
X['model'] = X['model'].apply(lambda x: apply_regex_rules(x) if isinstance(x, str) else x)

# unique models in the column model
unique_models = X['model'].unique()
print("\nUnique models in 'model':\n", unique_models)
# print the number of unique models
print("\nNumber of unique models in 'model':", len(unique_models))

In [ ]:
# Count all the entries in the json to compare with the unique models in the column model
with open("../mapping/modelname_mapping.json", "r", encoding="utf-8") as f:
    MM = json.load(f)
ALIASES = MM.get("aliases", {})
print("\nNumber of entries in the modelname_mapping.json:", len(ALIASES))

- Now we have all the modelnames fixed and normalized. We can use this to fill the missing values in the column Brand.

In [ ]:
# Check for missing values in 'Brand' before filling
print("\nMissing values in 'Brand' before filling:\n", X['Brand'].isnull().sum())

# Load mapping
with open("../mapping/brand_model_mapping.json", "r", encoding="utf-8") as f:
    brand_to_model = json.load(f)

# Map directly using the mapping
X["Brand"] = X["Brand"].fillna(X["model"].map(brand_to_model))

# Check for missing values in 'Brand' after filling
print("\nMissing values in 'Brand' after filling:\n", X['Brand'].isnull().sum())

In [ ]:
# Check for nan_brand values with models that could not be mapped
nan_brand_models = X[X['Brand'].isnull()]['model'].unique()
print("\nModels with missing 'Brand' after mapping:\n", nan_brand_models)

- We have mapped every possible model to its brand. The remaining missing values in the column Brand are due to missing model names.
  

In [ ]:
# Look at the Missing values in x and y
missing_values_x = X.isnull().sum()
missing_values_y = y.isnull().sum()

print("\nMissing values in X:\n", missing_values_x[missing_values_x > 0])


In [ ]:
# Show the entries with missing values in Brand
missing_brand_entries = X[X['Brand'].isnull()]
display(missing_brand_entries)

In [ ]:
# Show the entries with missing values in model
missing_model_entries = X[X['model'].isnull()]
display(missing_model_entries)

- Here we should decide to delete the 150 Entries with missing Brand and Model. 
- But we will keep them for now and focus on the next column 'year'.
- I will try to predict the missing model based on the other columns later, when the other columns are cleaned.

In [ ]:
# Show the percentage of the entries which have a missing brand and a missing model
missing_brand_model_percentage = (X['Brand'].isnull() & X['model'].isnull()).mean() * 100
print(f"\nPercentage of entries with missing 'Brand' and missing 'model': {missing_brand_model_percentage:.2f}%")

In [ ]:
# Delete the entries missing_brand_model_percentage from X
# print the shape of X before deleting
print("\nShape of X before deleting entries with missing 'Brand' and 'model':", X.shape)
# delete the entries    
X = X[~(X['Brand'].isnull() & X['model'].isnull())].reset_index(drop=True)
print("\nShape of X after deleting entries with missing 'Brand' and 'model':", X.shape)

In [ ]:
# Look at the Missing values in x and y
missing_values_x = X.isnull().sum()

print("\nMissing values in X:\n", missing_values_x[missing_values_x > 0])


- We will fill the value of year with the median within brand/model.
- But first we have to fix the wrong values in the column year.

In [ ]:
# Plot the distribution of the year column
plt.figure(figsize=(10, 6))
sns.histplot(X['year'].dropna(), bins=30, kde=True)
plt.title('Distribution of Year')
plt.xlabel('Year')
plt.ylabel('Frequency')
plt.show()

In [7]:
# print all values before 1995
print(X[X['year'] < 1995])

       carID     Brand     model    year transmission  mileage fuelType  \
11425  62732      OPEL    Zafira  1970.0       Manual  37357.0      NaN   
34917  35769  Mercedes   M Class  1970.0    Automatic  14000.0   Diesel   

         tax   mpg  engineSize  paintQuality%  previousOwners  hasDamage  
11425  200.0  42.2         NaN           60.0             2.0        0.0  
34917  305.0  39.2         0.0           44.0             3.0        0.0  


- There are 2 cars as outliers with the year 1970, so we will delete them, as they are too old and would effect our predictions negatively.

In [ ]:
# Delete the entries with year 1970
X = X[X['year'] != 1970]

# Check if the entries with year 1970 are deleted
print(X[X['year'] == 1970])

In [ ]:
# Boxplot the year column to see the outliers
plt.figure(figsize=(10, 6))
sns.boxplot(x=X['year'])
plt.title('Boxplot of Year')
plt.xlabel('Year')
plt.show()

In [ ]:
# Count the outliers outside the boxplot whiskers
Q1 = X['year'].quantile(0.25)
Q3 = X['year'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
outliers = X[(X['year'] < lower_bound) | (X['year'] > upper_bound)]
print(f"\nNumber of outliers in 'year': {outliers.shape[0]}")

In [ ]:
# Show some entries after 2020 in a dataframe
future_years = X[X['year'] > 2020]
display(future_years)

- It should be possible to have cars after 2020 because the dataset is from 2020. So the only option is to delete them or to replace them with the median of the respective brand/model.

In [ ]:
# Check for nan values in year column
print("\nMissing values in 'year':\n", X['year'].isnull().sum())

In [ ]:
# print the entries with nan values in year
nan_year_entries = X[X['year'].isnull()]
display(nan_year_entries)

In [ ]:
# Print the datatype of year column
print("\nData type of 'year':", X['year'].dtype)

- So we have a few entries with year > 2020 and we have some missing values in year.
- We will replace the wrong and missing values with the median of the respective brand/model.
- We will also convert the year column to integer type.

In [ ]:
import pandas as pd
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import KNNImputer

# Columns to use
cols = ['Brand', 'model', 'mileage', 'year', 'previousOwners']
df_subset = X[cols].copy()

# We'll only impute where year is NaN or > 2020
mask_to_impute = df_subset['year'].isna() | (df_subset['year'] > 2020)

# Keep original year for safety
df_subset['year_original'] = df_subset['year']

# Define columns for preprocessing
cat_cols = ['Brand', 'model']
num_cols = ['mileage', 'year', 'previousOwners']  # 'year' included so KNN can impute it

preprocessor = ColumnTransformer([
    ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), cat_cols),
    ('num_scaled', StandardScaler(), ['mileage', 'previousOwners']),
    ('year_passthrough', 'passthrough', ['year'])
])

# KNN Imputation pipeline
knn_pipeline = Pipeline([
    ('preprocess', preprocessor),
    ('impute', KNNImputer(n_neighbors=5, weights='distance'))
])

# Fit on the whole subset (KNN needs neighbors), then transform
imputed_array = knn_pipeline.fit_transform(df_subset[cat_cols + num_cols])

# The imputed 'year' is the LAST column in the transformed numeric block
imputed_years = imputed_array[:, -1]

# Build result and only overwrite where desired
df_imputed = df_subset.copy()
df_imputed.loc[mask_to_impute, 'year'] = imputed_years[mask_to_impute.values]

# Postprocess/guards
df_imputed['year'] = (
    df_imputed['year']
      .clip(lower=1980, upper=2020)
      .round()
      .astype(int)
)

# Optional: add a flag for transparency/debugging
df_imputed['year_imputed_knn'] = mask_to_impute.astype(int)

print(df_imputed[['Brand','model','mileage','previousOwners','year_original','year_imputed_knn','year']].head())


In [ ]:
# df_imputed only print the ones where year_original is nan or > 2020 as a dataframe
df_imputed_filtered = df_imputed[df_imputed['year_original'].isna() | (df_imputed['year_original'] > 2020)]
# display the dataframe as table
display(df_imputed_filtered)

In [ ]:
# Lets make a boxplot for every column in X in a dashboard to see the outliers
import matplotlib.pyplot as plt

plt.figure(figsize=(15, 10))
for i, col in enumerate(X.columns):
    # Only plot numeric columns, ignore non-numeric
    numeric_cols = X.select_dtypes(include=[np.number]).columns
    for i, col in enumerate(numeric_cols):
        plt.subplot(3, 3, i + 1)
        plt.boxplot(X[col].dropna())
        plt.title(col)
    break  # Prevents repeating for each column in X.columns
    numeric_cols = X.select_dtypes(include=[np.number]).columns
    for i, col in enumerate(numeric_cols):
        plt.subplot(3, 3, i + 1)
        plt.boxplot(X[col].dropna())
        plt.title(col)
    plt.title(col)
plt.tight_layout()
plt.show()

In [ ]:
# Show df with previous owner less 0

df_previous_owners_invalid = X[X['previousOwners'] < 0]
display(df_previous_owners_invalid)

In [ ]:
# Show datatype of X
print("\nData types of X:\n", X.dtypes)

- Previous Owners need to be integer
- year need to be integer too

In [ ]:
# Previous owners more than 5
df_previous_owners_high = X[X['previousOwners'] > 6]
display(df_previous_owners_high)

In [ ]:
# Show df with entries paintQuality% over 100%
df_paint_quality_invalid = X[X['paintQuality%'] > 100]
display(df_paint_quality_invalid)


In [ ]:
# Invalid Tax negative values
df_tax_invalid = X[X['tax'] < 0]
display(df_tax_invalid)

In [ ]:
# Invalid Tax values: 0 tax but everything else than electric
df_tax_invalid_0 = X[(X['tax'] == 0) & (X['fuelType'] != 'Electric')]
display(df_tax_invalid_0)

In [ ]:
# Cars with negative mileage
df_negative_mileage = X[X['mileage'] < 0]
display(df_negative_mileage)

- We can try to calculate the CO2 Emissions from the engineSize, fueltype, Mpgh, year and price. With the CO2 Emissions we can calculate the tax and see if they are correct.

In [ ]:
#Show me the cars with engineSize over 5L

df_large_engine = X[X['engineSize'] > 5.0]
display(df_large_engine)

# And the unique models with engineSize over 5L
unique_large_engine_models = df_large_engine['model'].unique()
print("\nUnique models with engineSize over 5L:\n", unique_large_engine_models)

In [ ]:
# Engine size less than 1 L
df_small_engine = X[X['engineSize'] < 1.0]
display(df_small_engine)

In [ ]:
# Show all unique values of transmission
unique_transmissions = X['transmission'].unique()
print("\nUnique values in 'transmission':\n", unique_transmissions)

In [ ]:
import pandas as pd
import numpy as np
import re

# Define a normalization function
def clean_transmission(x):
    if pd.isna(x):
        return np.nan
    s = str(x).strip().lower()

    # handle common patterns
    if re.search(r"man", s):          # e.g. 'manual', 'manua', 'manuaL '
        return "manual"
    elif re.search(r"semi", s):       # e.g. 'semi-auto', 'semi-aut'
        return "semi-auto"
    elif re.search(r"auto", s):       # e.g. 'automatic', 'automati', 'utomatic'
        return "automatic"
    elif re.search(r"unk", s):        # e.g. 'unknown', 'unknow', 'nknown'
        return "unknown"
    elif re.search(r"oth", s):        # e.g. 'other'
        return "other"
    else:
        return np.nan   # if none matches, mark as missing

# Apply it
X["transmission"] = X["transmission"].apply(clean_transmission)

# Optional: check results
print(X["transmission"].value_counts(dropna=False))

In [ ]:
# Show 10 entries of others in transmission
df_others_transmission = X[X['transmission'] == 'unknown']
display(df_others_transmission)

In [ ]:
# Create a JSON with all the unique brand and model which have an unknown transmission or are nan in transmission
df_unknown_transmission = X[X['transmission'].isin(['unknown', np.nan])]
# Brand,model, fuelType, engineSize, year
unique_unknown_transmission = df_unknown_transmission[['Brand', 'model', 'fuelType', 'engineSize']].drop_duplicates()
unique_list = unique_unknown_transmission.to_dict(orient='records')
with open("../mapping/unknown_transmission_models.json", "w", encoding="utf-8") as f:
    json.dump(unique_list, f, indent=4)

#print the number of unique models with unknown or nan transmission
print("\nNumber of unique models with unknown or nan transmission:", len(unique_unknown_transmission))


- Try to predict the transmission based on the other columns.

In [ ]:
# Show all unique values of fuelType
unique_fuel_types = X['fuelType'].unique()
print("\nUnique values in 'fuelType':\n", unique_fuel_types)

In [ ]:

def clean_fuel_type(x):
    if pd.isna(x):
        return np.nan
    s = str(x).strip().lower()

    # Match the main types
    if re.search(r"pet", s):          # petrol, petro, etrol, etc.
        return "petrol"
    elif re.search(r"die", s):        # diesel, diese, iese, etc.
        return "diesel"
    elif re.search(r"hyb", s):        # hybrid, hybr, ybrid, etc.
        return "hybrid"
    elif re.search(r"elec", s):       # electric
        return "electric"
    elif re.search(r"oth", s):        # other, othe, ther
        return "other"
    elif re.search(r"unk", s):        # if any 'unknown' appears
        return "unknown"
    else:
        return np.nan   # not recognized

# Apply cleaning
X["fuelType"] = X["fuelType"].apply(clean_fuel_type)
# Check results
print(X["fuelType"].value_counts(dropna=False))


In [ ]:
# X.head()
X.head(10)

In [ ]:
# unique values of hasDamage
unique_has_damage = X['hasDamage'].unique()
print("\nUnique values in 'hasDamage':\n", unique_has_damage)

In [ ]:
# count the number of True and False in hasDamage
has_damage_counts = X['hasDamage'].value_counts(dropna=False)
print("\nCounts of values in 'hasDamage':\n", has_damage_counts)

In [ ]:
# Check if all carID in X are unique
all_unique_carID = X['carID'].is_unique
print("\nAre all 'carID' values unique?:", all_unique_carID)